### ISSUE TITLE
Design Proposal - Race Traits

### ISSUE THREAD

**Issue by BottledByte** [tuomount/Open-Realms-of-Stars#673]
Races could be described by traits, both physical and mental. Such change would allow behavior (code) reuse, as any race with specified traits (or lack of traits) can be controlled by single code, which then does not have to be updated for each race. Such design can also be easily moved to data files (as per GH-666).

`RaceTrait` **could work as a flag with extra information**, like description and list of related traits (conflicting and/or required).
**It should NOT be implemented as enum**, however, because that would harm future extensibility and data-driven benefits.

Race trait system has following benefits over current design:
1. *Code reuse* - Trait can define behavior, not only race itself.
2. **Easier content addition** - Race is composed from traits,
   meaning less code has to be written or examined when doing changes.
3. *Data-driven design* - Since traits are data, this data can be de/serialized and tweaked without modifying code.
3. *Trait dependencies* - Some traits might require absence or presence of other traits to be applicable.
4. *Trait scoring/type* - Traits can have types like "good", "bad", "neutral", etc.
   and/or numeric scoring of how "good" they are. This can be reused by AI or in content generation.
5. *Better visualization* - Allows distinctive visualization in GUI,
   making it more "rich" and/or informative.
6. **Opportunity for user-made-content** - Players could theoretically compose a race from traits,
   creating new unique races, and thus play-styles, or customizing races for their play-style.
7. **Opportunity for content procedural generation** - Race could be composed from
   traits even at runtime, by the game, creating new unique races/games.

---

## Programmer Point of View

As a programmer, I hate duplicate, incomprehensible code, with many hard-to-track edge cases. Racial traits promote code reuse, (slightly) improve encapsulation and allow me to create new races using composition, reducing cognitive overhead and making my work way easier.

## Player Point of View

As a player, I want 4X strategy game with a lot of content and replayability. Racial traits offer this to me, as I could experience more unique games and even customize the game more.

I also like to clearly see what each game aspect/feature/entity does. Race traits clearly describe the race's abilities, maybe even graphically (like with icons), so I am informed adequately.

---

## Notes

The content generation opportunities are just opportunities such redesign offers. *There is no need to use those newfound abilities in game design.*

**I however highly recommend using those opportunities in game design, since it would bring massive amount of value to the player.**

During designing and prototyping this, I also discovered several logical inconsistencies and probably bugs in game code regarding races and their "traits" (which are not a thing now). I will add partial enumeration of such issues with explanations and other insights in next post.

## Summary

Adding and using race trait system is the first step of limiting impact of `SpaceRace` enum, making it easier to be *eventually* de-hardcoded as part of GH-666, while adding opportunities for content-generation and moddability, as well as fixing and preventing future bugs due to ad-hoc copy-paste and having to tweak tens of lines of code with addition of a race. It will also remove ambiguity of race-related code.

I have changes implementing the core of this ready, awaiting review and merge of my refactors as I depend on them. :+1:

**Comment by tuomount** [comment]
I totally agree this should be done, since it would allow to have custom races. This is quite big change.

### Recognized inconsistency 1
This is obvious bug. Lithovorians leaders would not benefit from agricultural perk.

### Recognized inconsistency 2
Mechions are creatures which have high capacity energy sources built-in and those does not require replacement.
Reason why there is credit cost after 4 Mechion population is that they require a bit of maintenance to keep them working. This is for balancing reasons. But it is true it is not mentioned anywhere. Could can think that each population has 0.25 credit of maintenance cost.

###  Recognized inconsistency 3
Again bug. Synthdroids were done after Deadly virus event and I missed that news event. But yes having racial traits would be good. How ever not sure how that would affect or how easy it is to get working AI for planet handling based on traits.

**Comment by tuomount** [comment]
> First of all, yes, attitude should be cut-of from SpaceRace, at least in it's current form. It is already very confusing with all the attitude "sources". #624 is nice example of how confusing it can be 🙂
> 
> Second, you are answering your own questions yourself. Since it is theoretically possible to derive attitude from race's traits/attributes, why not do it? And also, there is nothing wrong about random attitudes. Again, it adds variety. And Reborgians would make great diplomatic traders. "You give Planet, we go away." 😄 Yes, even this bizarre scenario with race practically designed for conquest trying to act like a diplomat and miserably fail is a gameplay.

Something like that actually happened me in one of the test plays I had. Reborgians were elder race, when I first met them, they wanted to trade map of planets and they even sweetened deal giving something extra tech for me. I agreed and then next meeting, they announced now you will die. :smile: 
Diplomacy bonus or malus is guidance for AI how easily it is willing to do deal. Negative means that another party requires something extra on deal, positive means that deals will be a bit cheaper.

> So, in the end, I see a system where attitudes are scored and weighted by "how they fit for the race", then one of them is randomly selected, and _assigned to AI_ (human player should not use attitude at all). I haven't added weighted lists, because I wanted to fight with Java's generics. 😄 Such system make most games normal, but kept things like "friendly Reborgians from your neighborhood" appearing on occasion.
> 

> Indeed, a mechanic where some realm has ships that are stealthy from game start is good. And it is not subject of removal. It is more like that "Has cloaked ships" trait will be renamed to "Stealthy" 😄. So that it matches design logic. It can be said that given the race is naturally stealthy, it also keeps their ships in "low-power" mode or similar, to simulate their own properties in space (since they are used to have stealthy advantage). _Trait justified._

Calling it stealthy is fine.

> > But yes there could be trait for better spy which would give bonus for espionage missions.
> 
> I would add that to the Stealthy trait, as per reasoning described above and in previous conversation regarding "chameleon-like abilities". Both bonuses to cloaking and espionage _missions_ (for leaders, as they are stealthy) should be rather small, like +5 or something. This is an example where small number makes large difference in balancing (in terms of cloaking). The espionage bonus is a "nice to have", that could incentivize recruitment of Stealthy-race leaders as spies. Which can consequently benefit game's "leader trading" mechanic.
> 
> And the **realm espionage** mechanic should be renamed to **intelligence**. Because there are two espionages, one being about realm-wide data (like ship positions, etc), while the second is mostly unrelated mechanic of **leader espionage missions**. It is confusing, both in-code and in-game. I will open a separate issue for that.
 
Calling realm espionage as intelligence is fine and makes sense since now these two are called almost same names.

> TLDR: The "cloaking device trait" will not be removed, just renamed and slightly _extended_.
> 
> > Charming and repulsive I was planning to add give diplomatic bonus. Humans should have diplomatic or similar trait which would give +2, charming +1, repulsive -2.
> 
> Another design thing: **Don't create many "insignificant levels of traits" for one basic aspect.** Two levels both ways are absolute maximum (like -2, -1, 0, +1, +2). After that, it becomes hard for player to keep track of traits as "categorization marker". Too high granularity for one aspect is bad for traits. Think about it like this:
> 
> As a player, I check out a race in intelligence view that I just met.
> 
> * With traits, I look and see "Ah, they are Strong, so they will have stronger troops than I have, since I am not Strong."
> * With a full numeric system, I can think "Ah, they have 13 troop power, I have 12, they are stronger, but not that much".
> * With granulated traits, I see "Ah, they are Mildly Strong, but I am Slightly Weak... They are stronger, but how big is the difference?"
How about there would be numeric system in code, but only showing textual information to player. Numeric value could be shown when conquering then calculate actual troop power or similarly in planet view show troop power somewhere. In that case player would get feedback if barracks or such is being built and would also know how long bomb the planet before sending the troops.

> So the 2 levels both directions (bonus/malus) are maximum in terms of player's cognition. And the first level should be an above average in that case, while the second level should be very dramatic. In terms of diplomacy, I will have 3 traits specific for that aspect:
> 
> * Natural Charm (+1) => +2 diplomacy
> * Repulsive (-1) => -2
> * Disgusting (-3) => -7
> 
> Trait points valuation in brackets. Notice the gaps. Additionally, there can be some other trait, that adds very minor (+1/-1) bonus to diplomacy, while focusing on other aspect or just describing something, but those should be both rare and "not game-changers in terms of diplomacy".
>  
> TLDR: Rushing mechanic should be decoupled from races completely.

Yes, this sounds good idea.

> How do you justify the traits for rushing with credits/population? How is rushing related to race? What is even rushing? I know you have added the traits to keep the existing systems working, but there are unanswered questions. **If population rush is slavery, then it is not a racial trait!**. Governmental aspect? For sure, but not racial aspect.

This is more governmental aspect.

> It could understand Population rush trait for i.e. Robotic races, which could "deconstruct/transform themselves" to aid building the project, or similar _weak explanation_. But its really weak. And the Credit rushing RaceTrait... I cannot explain that one _at all_.
> 
> Again, race physique, mentality and realm governance is mixed-up here. It makes the design smell. And it makes game confusing. I personally was not able to understand "why some of those rushing buttons are sometimes greyed-out when I play for different realm". It can be from Government, it can be from Race... and in case of Sporks, they even had special ability where both rushing methods were allowed 🙃 . _VERY confusing._
Rushing options should be only game from Governments. Let's change that.

**Comment by tuomount** [comment]
Would it help, if there would be another enum class for example SpaceRaceEnum which would contain those spacerace enums and include for example theme music, ship background, ship image, race picture etc. Then current SpaceRace would be turn into class which would contain Race name, which images use for ship, race, bridge and so on. These could get value from spacerace enums, but still use actual file paths. Reason why to use that SpaceRaceEnum is that when there is custom race, it would be nice to have some kind of name for example ship style and so on. Of course those images should be also shown

**Comment by tuomount** [comment]
Works now. Closing.
